janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/StaticArrayEntryList.java

```diff
@@ -398,11 +401,89 @@ private static <E,D> EntryList of(Iterable<E> elements, StaticArrayEntry.GetColV
     private static <E,D> EntryList of(Iterator<E> elements, StaticArrayEntry.GetColVal<E,D> getter, StaticArrayEntry.DataHandler<D> dataHandler) {
         Preconditions.checkArgument(elements!=null && getter!=null && dataHandler!=null);
         if (!elements.hasNext()) return EMPTY_LIST;
+
+        EntryListComputationContext context = generateComputationContext();
+        applyElementsComputation(elements, getter, dataHandler, context);
+        return convert(context);
+    }
+
+    public static EntryListComputationContext generateComputationContext(){
         long[] limitAndValuePos = new long[10];
-        byte[] data = new byte[limitAndValuePos.length*15];
-        EntryMetaData[] metadataSchema = null;
-        int pos=0;
-        int offset=0;
+        return new EntryListComputationContext(limitAndValuePos, new byte[limitAndValuePos.length*15], null, 0, 0);
+    }
+
+    public static <E> void supplyEntryList(ChunkedJobDefinition<Iterator<E>, EntryListComputationContext, EntryList> chunkedJobDefinition,
+                                           StaticArrayEntry.GetColVal<E,StaticBuffer> getter,
+                                           ExecutorService executorService) {
+        supplyEntryList(chunkedJobDefinition, getter, StaticArrayEntry.StaticBufferHandler.INSTANCE, executorService);
+    }
+
+    private static <E,D> void supplyEntryList(ChunkedJobDefinition<Iterator<E>, EntryListComputationContext, EntryList> chunkedJobDefinition,
+                                              StaticArrayEntry.GetColVal<E,D> getter,
+                                              StaticArrayEntry.DataHandler<D> dataHandler,
+                                              ExecutorService executorService){
+        assert chunkedJobDefinition !=null && getter!=null && dataHandler!=null;
+
+        executorService.execute(() -> {
+
+            Queue<Iterator<E>> chunksQueue = chunkedJobDefinition.getDataChunks();
+
+            if(!chunkedJobDefinition.getProcessingLock().tryLock()){
+                return;
+            }
+
+            try {
+                if(chunkedJobDefinition.getResult().isDone()){
+                    return;
+                }
+
+                Iterator<E> elements = chunksQueue.isEmpty() ? Collections.emptyIterator() : chunksQueue.remove();
```

Comment:
Be careful: the producer does not lock the queue when adding new chunks. After you test emptiness here, the emptiness might have changed before you call `Collections.emptyIterator()` or `chunksQueue.remove()`. I am not sure if this could cause a bug or not.

Reply:
Currently, I only see one consumer (this thread) attempting to lock the processing lock. Did you intend to have the producer to use the lock but forgot to do so, OR there could actually be multiple consumers?

Reply:
> Be careful: the producer does not lock the queue when adding new chunks. After you test emptiness here, the emptiness might have changed before you call Collections.emptyIterator() or chunksQueue.remove(). I am not sure if this could cause a bug or not.

Yes, the producer of the queue just adds data without locking anything. It's expected. It's OK if `chunksQueue.isEmpty()` returns `true` here but later it becomes `false`. 
We are checking the queue again after we release the lock. Either this thread will re-run this whole block again, or CQL-IO thread (producer) will trigger the job which will re-run this block again. (the concept is explained [here](https://github.com/JanusGraph/janusgraph/pull/3760#discussion_r1207629701))

Reply:
> Currently, I only see one consumer (this thread) attempting to lock the processing lock. Did you intend to have the producer to use the lock but forgot to do so, OR there could actually be multiple consumers?

The producer doesn't need to lock anything. Producer's job is:
1. Add new data to the queue.
2. Ensure that either some thread is running this block OR enter this block otherwise.

There is always only one consumer of the data queue - this locked block.